sourcemap/store.go

```diff
@@ -45,38 +42,75 @@ var (
 // Store holds information necessary to fetch a sourcemap, either from an Elasticsearch instance or an internal cache.
 type Store struct {
 	cache   *gocache.Cache
-	esStore *esStore
+	backend backend
 	logger  *logp.Logger
+
+	sync.Mutex
+	inflight map[string]chan struct{}
+}
+
+type backend interface {
+	fetch(ctx context.Context, name, version, path string) (string, error)
 }
 
-// NewStore creates a new instance for fetching sourcemaps. The client and index parameters are needed to be able to
-// fetch sourcemaps from Elasticsearch. The expiration time is used for the internal cache.
-func NewStore(client elasticsearch.Client, index string, expiration time.Duration) (*Store, error) {
+func newStore(b backend, logger *logp.Logger, expiration time.Duration) (*Store, error) {
 	if expiration < 0 {
 		return nil, errInit
 	}
-	logger := logp.NewLogger(logs.Sourcemap)
+
 	return &Store{
-		cache:   gocache.New(expiration, cleanupInterval(expiration)),
-		esStore: &esStore{client: client, index: index, logger: logger},
-		logger:  logger,
+		cache:    gocache.New(expiration, cleanupInterval(expiration)),
+		backend:  b,
+		logger:   logger,
+		inflight: make(map[string]chan struct{}),
 	}, nil
 }
 
 // Fetch a sourcemap from the store.
 func (s *Store) Fetch(ctx context.Context, name string, version string, path string) (*sourcemap.Consumer, error) {
-	key := key([]string{name, version, path})
+	key := cacheKey([]string{name, version, path})
 
 	// fetch from cache
 	if val, found := s.cache.Get(key); found {
 		consumer, _ := val.(*sourcemap.Consumer)
 		return consumer, nil
 	}
 
+	// if the value hasn't been found, check to see if there's an inflight
+	// request to update the value.
+	s.Lock()
+	wait, ok := s.inflight[key]
+	if ok {
+		// found an inflight request, wait for it to complete.
+		s.Unlock()
+		<-wait
+		// Try to read the value again
+		// TODO: All waiting Fetch()'s will attempt to re-fetch once
+		// the coordinated backend request has finished. At some point
+		// we need to start failing incoming requests as more start
+		// piling up, if eg. the backend is down.
+		return s.Fetch(ctx, name, version, path)
```

Comment:
Wanted to call out this TODO. I have no idea what a reasonable value would be for limiting connections, if that's something we want to handle now.

Reply:
How about adding a timeout for waiting on the inflight backend request to return? If the timeout is exceeded, return an error for this request rather than calling `s.Fetch(..)` again. 
Restricting the number of allowed requests would be tricky IMO - it much depends on the setup (for example the rate limiter values).

Reply:
Now that the waiters semaphore is gone, I'm wondering if we need this timeout either. Seems to me like the caller should be in charge of how long it waits for a source map, via the provided context.

Reply:
I'm in favor of that